hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/test/TestAbstractSqlConnector.java

```diff
@@ -0,0 +1,327 @@
+            fields.add(new MappingField(names[i], resolveTypeForTypeFamily(QueryDataTypeFamily.valueOf(types[i]))));
+        }
+        return fields;
+    }
+
+    @Nonnull
+    @Override
+    public Table createTable(
+            @Nonnull NodeEngine nodeEngine,
+            @Nonnull String schemaName,
+            @Nonnull String mappingName,
+            @Nonnull String externalName,
+            @Nonnull Map<String, String> options,
+            @Nonnull List<MappingField> resolvedFields
+    ) {
+        String[] names = options.get(OPTION_NAMES).split(DELIMITER);
+        String[] types = options.get(OPTION_TYPES).split(DELIMITER);
+
+        assert names.length == types.length;
+
+        List<TableField> fields = new ArrayList<>(names.length);
+        for (int i = 0; i < names.length; i++) {
+            fields.add(new TableField(names[i], resolveTypeForTypeFamily(QueryDataTypeFamily.valueOf(types[i])), false));
+        }
+
+        List<Object[]> rows = new ArrayList<>();
+        String[] rowsSerialized = options.get(OPTION_VALUES).split(VALUES_DELIMITER);
+        for (String rowSerialized : rowsSerialized) {
+            if (rowSerialized.isEmpty()) {
+                continue;
+            }
+
+            String[] values = rowSerialized.split(DELIMITER);
+
+            assert values.length == fields.size();
+
+            Object[] row = new Object[values.length];
+            for (int i = 0; i < values.length; i++) {
+                String value = values[i];
+                if (NULL.equals(value)) {
+                    row[i] = null;
+                } else {
+                    row[i] = fields.get(i).getType().convert(values[i]);
+                }
+            }
+            rows.add(row);
+        }
+
+        boolean streaming = Boolean.parseBoolean(options.get(OPTION_STREAMING));
+        return new TestBatchTable(this, schemaName, mappingName, fields, rows, streaming);
+    }
+
+    @Nonnull
+    @Override
+    public Vertex fullScanReader(
+            @Nonnull DAG dag,
+            @Nonnull Table table_,
+            @Nullable Expression<Boolean> predicate,
+            @Nonnull List<Expression<?>> projection,
+            @Nullable FunctionEx<ExpressionEvalContext, EventTimePolicy<Object[]>> eventTimePolicyProvider
+    ) {
+        EventTimePolicy<Object[]> eventTimePolicy = eventTimePolicyProvider == null
+                ? EventTimePolicy.noEventTime()
+                : eventTimePolicyProvider.apply(null);
+
+        TestBatchTable table = (TestBatchTable) table_;
+        List<Object[]> rows = table.rows;
+        boolean streaming = table.streaming;
+
+        FunctionEx<Context, TestBatchDataGenerator> createContextFn = ctx -> {
+            ExpressionEvalContext evalContext = SimpleExpressionEvalContext.from(ctx);
+            return new TestBatchDataGenerator(rows, predicate, projection, evalContext, streaming);
+        };
+
+        ProcessorMetaSupplier pms;
+        if (table.streaming) {
```

Comment:
This class is abstract yet it contains logic belonging to its implementors... maybe get rid of `table.streaming` and move implementor specific logic to respective connectors?

Reply:
I moved some code down, but not all. This way it's less code.